test/jdk/java/io/Reader/Of.java

```diff
@@ -0,0 +1,217 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.Reader;
+import java.io.StringReader;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.ReadOnlyBufferException;
+
+import org.testng.annotations.*;
+
+import static org.testng.Assert.*;
+
+/*
+ * @test
+ * @bug 8341566
+ * @summary Check for expected behavior of Reader.of().
+ * @run testng Of
+ */
+public class Of {
+    final static String CONTENT = "Some Reader Test";
+
+    /*
+     * Readers to be tested.
+     */
+    @DataProvider
+    public static Reader[] readers() {
+        return new Reader[] {
+            new StringReader(CONTENT),
+            Reader.of(CONTENT),
+            Reader.of(new StringBuffer(CONTENT)),
+            Reader.of(new StringBuilder(CONTENT)),
+            Reader.of(ByteBuffer.allocateDirect(CONTENT.length() * 2)
+                    .asCharBuffer().put(CONTENT).flip()),
+            Reader.of(CharBuffer.wrap(CONTENT.toCharArray())),
+            Reader.of(new CharSequence() {
+                @Override
+                public char charAt(int index) {
+                    return CONTENT.charAt(index);
+                }
+
+                @Override
+                public int length() {
+                    return CONTENT.length();
+                }
+
+                @Override
+                public CharSequence subSequence(int start, int end) {
+                    // unused by Reader.Of's result
+                    throw new UnsupportedOperationException();
+                }
+
+                @Override
+                public String toString() {
+                    // Reader.Of's result SHALL NOT convert to String
+                    throw new UnsupportedOperationException();
+                }
+            })
+        };
+    }
+
+    @Test(dataProvider = "readers")
+    public void testRead(Reader reader) throws IOException {
+        String s = "";
+        for (int c; (c = reader.read()) != -1; s += (char) c);
+        assertEquals(s, CONTENT, "read() returned wrong value");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadBII(Reader reader) throws IOException {
```

Comment:
What does `BII` in some of these test method names represent?

Reply:
Good question. I hereby forward it to the original author of the existing tests I have copied from OpenJDK. I assume it was some kind of convention.